orttraining/orttraining/python/training/ortmodule/_logger.py

```diff
@@ -23,31 +26,71 @@ class LogLevel(IntEnum):
 
 
 @contextmanager
-def suppress_os_stream_output(suppress_stdout=True, suppress_stderr=True, log_level=LogLevel.WARNING):
-    """Suppress output from being printed to stdout and stderr if log_level is WARNING or higher.
+def suppress_os_stream_output(on_exit: Optional[Callable] = None):
+    """Suppress output from being printed to stdout and stderr.
 
-    If there is any output detected, a single warning is issued in the context
+    If on_exit is not None, it will be called when the context manager exits.
     """
 
     # stdout and stderr is written to a tempfile instead
-    stdout = sys.stdout
-    stderr = sys.stderr
+    _, fo_file_path = tempfile.mkstemp()
 
-    suppress_logs = log_level >= LogLevel.WARNING
+    # Store original stdout and stderr file no.
+    old_stdout = os.dup(sys.stdout.fileno())
+    old_stderr = os.dup(sys.stderr.fileno())
 
-    fo = io.StringIO()
+    # Allow read and write the file.
+    fd = os.open(fo_file_path, os.O_RDWR | os.O_CREAT)
+    fo = os.fdopen(fd)
 
     try:
-        if suppress_stdout and suppress_logs:
-            sys.stdout = fo
-        if suppress_stderr and suppress_logs:
-            sys.stderr = fo
-        yield fo
+        # Redirect stdout and stderr (printed from Python or C++) to the file.
+        os.dup2(fo.fileno(), sys.stdout.fileno())
+        os.dup2(fo.fileno(), sys.stderr.fileno())
+        yield
     finally:
-        if suppress_stdout:
-            sys.stdout = stdout
-        if suppress_stderr:
-            sys.stderr = stderr
+        sys.stdout.flush()
+        sys.stderr.flush()
+
+        # Restore stdout and stderr.
+        os.dup2(old_stdout, sys.stdout.fileno())
+        os.dup2(old_stderr, sys.stderr.fileno())
+
+        if on_exit:
+            on_exit(fo)
+
+        fo.close()
+
+
+def create_log_filter(logger: logging.Logger, record_filters: Optional[List[str]], name: Optional[str]):
+    def _log_with_filter(fo):
+        if fo.tell() > 0:
+            if logger.disabled:
+                return
+            # Filter out the error message
+            fo.seek(0)
+            suppress_output_messages = fo.readlines()
+            if record_filters:
+                filtered_messages = []
+                is_filtered = False
+                for suppressed_message in suppress_output_messages:
```

Comment:
String matching can take a long time (specially in python).

I once found that pytorch C++ log filtering added a 2~2.5% penalty to ORTModule runs. Hoping that the filtering here will not have a negative impact on the performance.